### PR TITLE
Implement ID algorithm for causal effect identification

### LINKED ISSUE #3465: [ENH] Implement ID algorithm for causal identification in semi-Markovian models
**Is your feature request related to a problem? Please describe.**
pgmpy's existing identification methods (backdoor, frontdoor, IV) are incomplete they fail on valid causal models where none of the role-based criteria apply. The canonical example is the bow-arc graph `(X -> Z -> Y, X <-> Z)`, where `P(Y | do(X))` is identifiable but no existing pgmpy method can find it. The ID algorithm (Shpitser & Pearl, 2006) is complete for semi-Markovian models and strictly subsumes all three existing methods.

**Describe the solution you'd like**
Add `pgmpy/identification/id.py` with:
`_id_recursive(y, x, P, G)`: module-level function implementing the nine-rule recursion. It must be a standalone function rather than a method on ID because recursive sub-calls operate on internal subgraphs that carry no role assignments the class-level validation in `_validate_query` would break these calls.

`ID(_BaseFormulaIdentification):`

- `supported_graph_types = (ADMG, DAG)`
- `_identify(causal_graph)`: extracts y and x from roles, initialises `P = Prob(frozenset(causal_graph.nodes()))`, calls `_id_recursive`, wraps the result in `ProbabilityExpression`. On failure, stores the witness subgraph in `self.hedge_` and returns `False`.

Add tests for the ID algorithm a few of them are listed below :

- Bow-arc graph: ID succeeds and returns the correct formula; backdoor and frontdoor both fail on the same graph.
- Simple DAG with no confounding: reduces to the expected direct observational expression.
- Frontdoor graph: recovers the marginal-over-mediator form.
- Non-identifiable graph: returns False and self.hedge_ is a non-empty subgraph.

**Describe alternatives you've considered**

**Additional context**
Related Issue: #2529 

Reference proposal: [Expanding Causal Identification module](https://github.com/pgmpy/enhancement_proposals/tree/main/3_causal_identification).

[ID: Shpitser, I. and Pearl, J. (2006). Identification of Joint Interventional Distributions in Recursive Semi-Markovian Causal Models. AAAI-06.
](https://cdn.aaai.org/AAAI/2006/AAAI06-191.pdf)

### PR DESCRIPTION
**Pull request by kajal-jotwani**
**The following checklist is mandatory**

Your PR will be closed if you remove the checklist. Use of LLMs is **strictly forbidden** for any part of this checklist (including for improving language), and will result in a **ban** if we find any use of LLMs.

### Your checklist for this pull request

- [x] Have you followed all the steps from our [Contributing Guide](https://github.com/pgmpy/pgmpy/blob/dev/CONTRIBUTING.md)?
- [x] Does the PR fully address the linked issue and is within its defined scope? If you are still working on the PR, mark it as draft.
- [x] Are all the GitHub Actions checks passing? If not, mark your PR as draft while you fix it.

If you have used AI/LLMs for any assistance, please answer the following questions. Please refer [#2622](https://github.com/pgmpy/pgmpy/pull/2622) for an example of the level of detail we expect:

- [x] Have you reviewed our [AI usage policy](https://github.com/pgmpy/pgmpy/blob/dev/CONTRIBUTING.md#ai-usage-policy)?
- [x] Are you able to fully explain your changes? We expect you to fully understand the algorithm and take full responsibility for any changes in this PR.

- Please list the AI tool(s) used, along with the model and its version used.

[Please replace this with your answer]

- Please describe in **detail** how and what you used AI assistance for? Please outline your whole workflow with the AI tool.

[Please replace this with your answer]

- What steps have you taken to verify that the changes correctly address the issue? What edge cases have you considered? Other than running tests, what else have you verified?

[Please replace this with your answer]

- Have you used AI for generating tests? Can you compress them into a smaller number of tests without losing coverage?

[Please replace this with your answer]

### Issue number(s) that this pull request fixes
- Fixes #3465

### List of changes to the codebase in this pull request
-

### CODE REVIEW

File: pgmpy/identification/ID.py [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+    The class implements the recursive identification procedure of
+    :footcite:t:`shpitser_2006` for ADMGs with exposure and outcome roles.
+    When the effect is identifiable, it returns a symbolic probability
+    expression for the interventional distribution. When the effect is not
+    identifiable, it returns ``False`` and stores the witness hedge in
+    ``self.hedge_``.
+
+    Parameters
+    ----------
+    causal_graph : ADMG
```
Comment (DARHWOLF) [inline pgmpy/identification/ID.py]: can you update this to support DAGs too

File: pgmpy/identification/ID.py [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+    ProductNode,
+)
+
+
+class ID(BaseFormulaIdentification):
+    """
+    Given a causal graph, identifies the causal effect using the ID algorithm.
+
+    The class implements the recursive identification procedure of
+    :footcite:t:`shpitser_2006` for ADMGs with exposure and outcome roles.
```
Comment (DARHWOLF) [inline pgmpy/identification/ID.py]: ADMGs and DAGs

File: pgmpy/identification/ID.py [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+    Returns
+    -------
+    ProbabilityExpressionTree
+        The symbolic formula for P(Y|do(X)) when identifiable.
+    False
+        When the effect is not identifiable. The witness hedge is stored
+        in ``self.hedge_``.
+
+    Examples
+    --------
```
Comment (DARHWOLF) [inline pgmpy/identification/ID.py:31]: I think it would be nice to have some more examples...

File: pgmpy/identification/_base.py [hunk 4]
```diff
@@ -84,3 +84,132 @@ def validate(self, causal_graph):
...
+        ------
+        ValueError
+            If the graph type is not supported, or if a required role is
+            missing.
+        """
+        if not isinstance(causal_graph, self.supported_graph_types):
+            raise ValueError(
+                f"causal_graph must be an instance of "
+                f"{self.supported_graph_types} for this method. "
+                f"Got {type(causal_graph).__name__}."
```
Comment (ankurankan) [inline pgmpy/identification/_base.py]: Use 120 char line.

File: pgmpy/identification/_base.py
Comment (ankurankan) [inline pgmpy/identification/_base.py:1]: Go through the whole file and make all the lines 120 char long.

File: pgmpy/identification/ID.py [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+
+    Parameters
+    ----------
+    causal_graph : ADMG
+        An ADMG with exposures and outcomes roles assigned. The two roles must
+        be disjoint.
+
+    Returns
+    -------
+    ProbabilityExpressionTree
```
Comment (ankurankan) [inline pgmpy/identification/ID.py]: Returns format should be `variable name : type`

File: pgmpy/identification/ID.py [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+            The identified formula, or False if not identifiable.
+        """
+        # TODO: Remove this conversion once DAG inherits _CoreGraph.
+        if isinstance(causal_graph, DAG):
+            admg = ADMG(edge_list=[(u, v, "->") for u, v in causal_graph.edges()])
+            # `edge_list` alone silently drops nodes that have no edges.
+            admg.add_nodes_from(causal_graph.nodes())
+            for role, variables in causal_graph.get_role_dict().items():
+                admg.with_role(role=role, variables=variables, inplace=True)
+            causal_graph = admg
```
Comment (ankurankan) [inline pgmpy/identification/ID.py]: Create a new issue to write a generalized method in the `_CoreGraph` to be able to convert between graph types.

File: pgmpy/identification/ID.py [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+
+        exposures = frozenset(causal_graph.get_role("exposures"))
+        outcomes = frozenset(causal_graph.get_role("outcomes"))
+        variables = frozenset(causal_graph.nodes())
+
+        # Definition 2 of the paper defines P_x(Y) only for Y ∩ X = ∅.
+        if overlap := (exposures & outcomes):
+            raise ValueError(f"'exposures' and 'outcomes' must be disjoint, but {sorted(overlap)} is in both.")
+
+        # π, a single topological ordering fixed over the ORIGINAL graph and threaded
```
Comment (ankurankan) [inline pgmpy/identification/ID.py]: Do not use unicode characters.

File: pgmpy/identification/ID.py [hunk 8]
```diff
@@ -0,0 +1,268 @@
...
+        # TODO: Remove this conversion once DAG inherits _CoreGraph.
+        if isinstance(causal_graph, DAG):
+            admg = ADMG(edge_list=[(u, v, "->") for u, v in causal_graph.edges()])
+            # `edge_list` alone silently drops nodes that have no edges.
+            admg.add_nodes_from(causal_graph.nodes())
+            for role, variables in causal_graph.get_role_dict().items():
+                admg.with_role(role=role, variables=variables, inplace=True)
+            causal_graph = admg
+
+        exposures = frozenset(causal_graph.get_role("exposures"))
```
Comment (ankurankan) [inline pgmpy/identification/ID.py:78]: Instead of having long comments. Try to write comments as `Step x` which outlines the pseudo code of the algorithm.

File: pgmpy/identification/ID.py [hunk 9]
```diff
@@ -0,0 +1,268 @@
...
+        # P, the estimand. Starts as the observational joint P(v) and is rewritten
+        # by lines 2 and 7 as the recursion proceeds.
+        estimand = ProbabilityNode(variables)
+
+        result = self._identify_recursive(outcomes, exposures, variables, causal_graph, estimand, ordering)
+        if result is False:
+            return False
+        return ProbabilityExpressionTree(root=result)
+
+    def _marginalize(self, node, sumset):
```
Comment (ankurankan) [inline pgmpy/identification/ID.py]: Figure out what exactly this operation is. If it is something to do only with the expression, then it belongs to the ProbabilityExpression class.

File: pgmpy/identification/ID.py [hunk 10]
```diff
@@ -0,0 +1,268 @@
...
+            marginalised (line 2) or replaced by ``Q[S']`` (line 7).
+        ordering : list
+            ``π``, the topological ordering fixed over the original graph.
+
+        Returns
+        -------
+        _TreeNode or False
+            The expression for ``P_x(y)``, or False if a hedge was found.
+        """
+        # Line 1: x = ∅. The effect is a marginal of the carried estimand.
```
Comment (ankurankan) [inline pgmpy/identification/ID.py]: `Line` -> `Step`

**Review by ankurankan — CHANGES_REQUESTED** [review]

File: pgmpy/identification/ID.py [hunk 11]
```diff
@@ -0,0 +1,268 @@
...
+        The symbolic formula for P(Y|do(X)) when identifiable.
+    False
+        When the effect is not identifiable. The witness hedge is stored in
+        ``self.hedge_`` as the pair of graphs ``(G, G_S)`` thrown by line 5 of
+        the algorithm; the C-forests forming the hedge are edge subgraphs of
+        that pair (Theorem 6 of the paper).
+
+    Examples
+    --------
+    >>> from pgmpy.base import ADMG
```
Comment (DARHWOLF) [inline pgmpy/identification/ID.py:32]: would be nice to have more examples. Perhaps using DAGs?

File: pgmpy/identification/ID.py [hunk 12]
```diff
@@ -0,0 +1,268 @@
...
+            admg = ADMG(edge_list=[(u, v, "->") for u, v in causal_graph.edges()])
+            # `edge_list` alone silently drops nodes that have no edges.
+            admg.add_nodes_from(causal_graph.nodes())
+            for role, variables in causal_graph.get_role_dict().items():
+                admg.with_role(role=role, variables=variables, inplace=True)
+            causal_graph = admg
+
+        exposures = frozenset(causal_graph.get_role("exposures"))
+        outcomes = frozenset(causal_graph.get_role("outcomes"))
+        variables = frozenset(causal_graph.nodes())
```
Comment (DARHWOLF) [inline pgmpy/identification/ID.py:80]: Do we need to have a check here for at least one exposure/outcome being defined? I think the probability class probability expression class already checks for this.

File: pgmpy/tests/test_identification/test_id.py [hunk 13]
```diff
@@ -0,0 +1,328 @@
...
+        assert algorithm.identify(graph(BOW_ARC, {"X"}, {"Y"})) is False
+        assert algorithm.hedge_ is not None
+        assert algorithm.identify(graph(CHAIN, {"X"}, {"Y"})) is not False
+        assert algorithm.hedge_ is None
+
+
+class TestIDInputHandling:
+    def test_dag_input(self):
+        dag = DAG(ebunch=[("X", "M"), ("M", "Y")])
+        dag = dag.with_role("exposures", ["X"]).with_role("outcomes", ["Y"])
```
Comment (DARHWOLF) [inline pgmpy/tests/test_identification/test_id.py]: why not ust define roles the usual way?

File: pgmpy/identification/ID.py [hunk 14]
```diff
@@ -0,0 +1,268 @@
...
+
```
Comment (DARHWOLF) [inline pgmpy/identification/ID.py:81]: I do not understand how this handles latent variables from the DAG? Wouldn't this add all latent nodes from the DAG to variables? How will the ADMG convert to bidirected edges then to represent the latents?

**Review by DARHWOLF — COMMENTED** [review]
Nice work! This seems like quite a strong implementation. (Some clean up could help)

- I tried implementing the typical SCM to check whether this is handling latent variables from a DAG correctly while converting it to ADMG. I do not see it behaving corretly. 

- Please try implementing this test: 

```
from pgmpy.base import DAG
from pgmpy.identification import ID
dag = DAG(ebunch=[("U", "X"), ("U", "Y"), ("X", "Y")], latents={"U"})
dag = dag.with_role("exposures", ["X"]).with_role("outcomes", ["Y"])

result = ID().identify(dag)
assert result is False
```

I see an assertion error here, and the following output -- which i do not think is correct. ID should fail for this particular case:  

```
result.to_latex()

'\\sum_{U} P(U) P(Y \\mid U, X)'
```

File: pgmpy/identification/ID.py (hunk 12 above)
Comment (kajal-jotwani) [reply pgmpy/identification/ID.py:80]: No check is needed here as `_validate_causal_graph` calls `causal_graph.is_valid_causal_structure()` (pgmpy/base/_mixin_roles), which raises if either role is empty:
```
>>> ID().identify(DAG(ebunch=[("X", "Y")]))
ValueError: must have at least one 'exposures' and one 'outcomes' role defined 
```
> I think the probability class probability expression class already checks for this.

yes the `ProbabiltyNode` guards the empty var case too

File: pgmpy/identification/ID.py (hunk 14 above)
Comment (kajal-jotwani) [reply pgmpy/identification/ID.py:81]: I've fixed it now by converting the DAG to an ADMG before running the ID algorithm, so the latent variables don't get included in `variables` and are instead represented as bidirected edges.

I used [Verma's latent projection](https://www.jmlr.org/papers/volume9/shpitser08a/shpitser08a.pdf) as the reference for the conversion. Thanks for catching this!
